erc777.k

```diff
@@ -69,140 +72,379 @@ module ERC777
-       <operators> ... Holder |-> ( Operators => Operators -Set SetItem(Operator) ) ... </operators>
+  rule <k> revokeOperator ( Operator:Int ) => #void ... </k>
+       <caller> Holder </caller>
+       <id> Holder </id>
+       <operators> Operators => Operators -Set SetItem(Operator) </operators>
        <log> Log => Log RevokedOperator ( Operator , Holder ) </log>
        requires Holder =/=Int Operator
 
 /* FIXME: temporarily disable for quick testing
   // Non-deterministically reject to revoke a non-authorized-operator
-  rule <k> Holder : revokeOperator ( Operator ) => #revert ... </k>
+  rule <k> Holder : revokeOperator ( Operator ) => revert ... </k>
        <operators> ... Holder |-> Operators:Set ... </operators>
        requires Holder =/=Int Operator
         andBool Operator in Operators
 */
 
   // Reject to revoke itself -- an address must be an operator for itself
-  rule <k> Holder : revokeOperator ( Operator ) => #revert ... </k>
+  rule <k> revokeOperator ( Operator:Int ) => revert ... </k>
+       <caller> Holder </caller>
        requires Holder ==Int Operator
 
   //
   // send / operatorSend
-  //
+  // TODO: ERC20 compatibility
 
-  rule <k> Holder :         send (          To , Value , Data      )
-        => Holder : operatorSend ( Holder , To , Value , Data , "" ) ... </k>
+  rule <k>         send (          To:Int , Value:Int , Data      )
+        => operatorSend ( Holder , To , Value , Data , "" ) ... </k>
+       <caller> Holder </caller>
 
   ////
 
   // 'From == 0' implies 'From == Operator' -- TODO: what's the intuition? may be vulnerable
-  rule <k> Operator : operatorSend ( From     , To , Value , Data , OperatorData )
-        => Operator : operatorSend ( Operator , To , Value , Data , OperatorData ) ... </k>
+  rule <k> operatorSend ( From:Int , To:Int , Value:Int , Data , OperatorData )
+        => operatorSend ( Operator , To , Value , Data , OperatorData ) ... </k>
+       <caller> Operator </caller>
        requires From ==Int 0
 
-  rule <k> Operator : operatorSend ( From , To , Value , Data , OperatorData )
+  rule operatorSend ( From , To , Value , Data , OperatorData )
+       => #operatorSendAux ( From , To , Value , Data , OperatorData , #getBalance(From), #getBalance(To) )
+
+  syntax K ::= "#operatorSendAux" "(" Address "," Address "," Number "," Data "," Data "," AExp "," AExp ")" [seqstrict(6,7)]
+  rule <k> #operatorSendAux ( From , To , Value , Data , OperatorData , BalanceFrom:Int, BalanceTo:Int )
         => #atomicBegin
-        ~> #ensure ( #check ( Operator , From , To , Value , Data ) )
-//      ~> #hookFrom ( Operator , From , To , Value , Data , OperatorData )
+        ~> #ensure( #and(
+              isOperatorFor ( Operator , From ), #and(
+              To =/=Int 0,
+              Value >=Int 0,
+              BalanceFrom >=Int Value),
+              #or(From ==Int To, BalanceTo +Int Value <=Int #MAXVALUE),
+              #isMultipleOf ( Value , granularity() ) ))
```

Comment:
mixed feeling here. isn't it too complicated? i thought about it, but intentionally put some of them in the `#update`. let's discuss more.

btw, fix the nested `#and`.